core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -512,7 +560,89 @@ impl<Block: BlockT> Inner<Block> {
+				// let the local-view handle it and likely treat it as past or future.
+				return self.local_view.as_ref().map(|v| v.consider_vote(round, set_id))
+                    .unwrap_or(Consider::RejectOutOfScope)
+			}
+		};
+		let mut tally = tally_for_round.entry(msg.id.clone()).or_insert(Default::default());
+
+		// A tally of messages received per peer, per voter, per round.
+		let peer_tally_for_round = match self.incoming_msg_tally.get_mut(&round) {
+			Some(tally_for_round) =>  tally_for_round,
+			None => {
+				// We don't know about this round,
+				// let the local-view handle it and likely treat it as past or future.
+				return self.local_view.as_ref().map(|v| v.consider_vote(round, set_id))
+                    .unwrap_or(Consider::RejectOutOfScope)
+			}
+		};
+		let mut per_peer_tally = peer_tally_for_round.entry((msg.id.clone(), who.clone())).or_insert(Default::default());
+
+		// A tally of messages we sent out, per peer, per voter, per round
+		let outgoing_tally_for_round = self.outgoing_votes_tally
+			.get_mut(&round)
+			.expect("If incoming votes knows about this round, so should outgoing ones; qed");
+		let outgoing_tally = outgoing_tally_for_round.entry((msg.id.clone(), who.clone())).or_insert(Default::default());
+
+		match &msg.message {
+			PrimaryPropose(_propose) => {
+				if per_peer_tally.primary_proposals < 3 {
+					per_peer_tally.primary_proposals += 1;
+				}
+				if tally.primary_proposals < 2 {
+					tally.primary_proposals += 1;
+				}
+			},
+			Prevote(_prevote) => {
+				if per_peer_tally.pre_votes < 3 {
+					per_peer_tally.pre_votes += 1;
+				}
+				if tally.pre_votes < 2 {
+					tally.pre_votes += 1;
+				}
+			},
+			Precommit(_precommit) => {
+				if per_peer_tally.pre_commits < 3 {
+					per_peer_tally.pre_commits += 1;
+				}
+				if tally.pre_commits < 2 {
+					tally.pre_commits += 1;
+				}
+			},
+		}
+
+		// We reject all messages for a given voter,
+		// once we're received more than two messages of any kind for that voter.
+		//
+		// We report peers who send us a third or more of message of any kind,
+		// when we've already exchanged two messages of any kind for that voter to that peer.
+		let (should_reject, should_report) = {
+			let should_reject = tally.primary_proposals > 1 || tally.pre_votes > 1 || tally.pre_commits > 1;
+			let should_report = (outgoing_tally.primary_proposals + per_peer_tally.primary_proposals) > 2 ||
+				(outgoing_tally.pre_votes + per_peer_tally.pre_votes) > 2 ||
+				(outgoing_tally.pre_commits + per_peer_tally.pre_commits) > 2;
+			(should_reject, should_report)
+		};
+
+		// When a peer sends us a redundant message, and should have known better,
+		// ignore the message and report the peer.
+		if should_report {
+			return Consider::RejectWillfulRedundant
+		}
+
+		// When we're processed 2 votes of a given type,
+		// reject all further messages for this voter.
+		if should_reject {
+			return Consider::RejectRedundant
```

Comment:
I'm not convinced this is the responsibility of the networking system. IMO it only serves to complicate the logic here. GRANDPA internals will already reject the redundant votes, and although it does mean that we will have to do a little extra work to get them to GRANDPA in the first place, we know that the total number of messages to process will be bounded just because we bound per-peer comms.

Reply:
If we reject here they won't be stored by the networking layer. I thought that was a goal of this?

Reply:
The main goal is to make sure that the amount of data stored by the networking layer is bounded. This extra snippet means that we push the bound down a bit, but it doesn't give us an algorithmic win. (still O(n) in the number of peers).

Reply:
Ok, so just to be clear, we want to remove the `Consider::RejectRedundant`, but keep the `Consider::RejectWillfulRedundant`(reporting bad peers) one?

Reply:
Yeah. But I would merge it as is if there's a good argument for reducing the bound.

Reply:
I'd argue for removing it, if only because we have this equivocation api in the `Environment` and we might want to do something with it at some point, so it's probably a good idea to let those through to be handled downstream.